src/main/java/gregtech/common/items/behaviors/ColorSprayBehavior.java

```diff
@@ -106,9 +128,51 @@ private boolean tryPaintSpecialBlock(EntityPlayer player, World world, BlockPos
         return false;
     }
 
+    // note: automatically uses durability from recolouring pipes, apart from the last use, as this allows for proper
+    // animation of the item's use
+    private boolean traversePipes(EntityPlayer player, World world, EnumHand hand, BlockPos startPos,
+                                  EnumFacing facing) {
+        startPos = startPos.offset(facing);
+        TileEntity connectedTe = world.getTileEntity(startPos);
+        int count = 1;
+        boolean painted = false;
+        while (connectedTe instanceof IPipeTile<?, ?>connectedPipe && count < MAX_PIPE_TRAVERSAL_LENGTH) {
+            if (connectedPipe.getPaintingColor() != (this.color == null ? -1 : this.color.colorValue)) {
+                connectedPipe.setPaintingColor(this.color == null ? -1 : this.color.colorValue);
+                connectedPipe.scheduleRenderUpdate();
+                ItemStack heldItem = player.getHeldItem(hand);
+                if (getUsesLeft(heldItem) == 1 && !player.isCreative()) {
+                    useItemDurability(player, hand, heldItem, empty.copy());
+                    return true;
+                }
+                // Off by one durability as the final use is handled by onItemUse, along with the use animation
+                if (painted) {
+                    useItemDurability(player, hand, heldItem, empty.copy());
+                }
+                painted = true;
+            }
+            if (connectedPipe.getNumConnections() == 2) {
```

Comment:
i was initally basing this off how the GTNH one works, but this could work with a stack or a similar data structure, but i think thats too complex for what it is worth